src/manifolds/Tucker.jl

```diff
@@ -0,0 +1,741 @@
+
+    TuckerPoint(A :: AbstractArray{T, D}, mlrank :: NTuple{D, Int}) where {T, D}
+
+The low-multilinear rank tensor arising from the sequentially truncated the higher-order
+singular value decomposition of A [^Vannieuwenhoven2012].
+
+[^Vannieuwenhoven2012]:
+    > Nick Vannieuwenhoven, Raf Vandebril, Karl Meerbergen: "A new truncation strategy for the higher-order singular value decomposition"
+    > SIAM Journal on Scientific Computing, 34(2), pp. 1027-1052, 2012
+    > doi: [10.1137/110836067](https://doi.org/10.1137/110836067)
+
+"""
+struct TuckerPoint{T,D} <: AbstractManifoldPoint
+    hosvd::HOSVD{T,D}
+end
+function TuckerPoint(
+    core::AbstractArray{T,D}, factors::Vararg{MtxT,D}
+) where {T,D,MtxT<:AbstractMatrix{T}}
+    # Take the QR decompositions of the factors and multiply the R factors into the core
+    qrfacs = qr.(factors)
+    Q = map(qrfac -> qrfac.Q, qrfacs)
+    R = map(qrfac -> qrfac.R, qrfacs)
+    core′ = reshape(Kronecker.:⊗(reverse(R)...) * vec(core), size(core))
+
+    # Convert to HOSVD format by taking the HOSVD of the core
+    decomp = st_hosvd(core′)
+    factors′ = Q .* decomp.U
+    return TuckerPoint(HOSVD{T,D}(factors′, decomp.core, decomp.σ))
+end
+function TuckerPoint(A::AbstractArray{T,D}, mlrank::NTuple{D,Int}) where {T,D}
+    @assert is_valid_mlrank(size(A), mlrank)
+    return TuckerPoint(st_hosvd(A, mlrank))
+end
+
+@doc raw"""
+    TuckerTVector{T, D} <: TVector
+
+Tangent space to the Tucker manifold at $x = (U_1,\dots,U_D) ⋅ \mathcal{C}$. This vector is
+represented as
+```math
+(U_1,\dots,U_D) \cdot \dot{\mathcal{C}} +
+\sum_{d=1}^D (U_1,\dots,U_{d-1},\dot{U}_d,U_{d+1},\dots,U_D) \cdot \mathcal{C}
+```
+where $\dot{U}_d^\mathrm{H} U_d = 0$. See also [`Tucker`](@ref)
+"""
+struct TuckerTVector{T,D} <: TVector
+    Ċ::Array{T,D}
+    U̇::NTuple{D,Matrix{T}}
+end
+
+# An implicitly stored basis of the tangent space to the Tucker manifold. This is the basis
+# from [Dewaele2021] and acts as the default orthonormal basis.
+struct HOSVDBasis{T,D}
+    point::TuckerPoint{T,D}
+    U⊥::NTuple{D,Matrix{T}}
+end
+CachedHOSVDBasis{𝔽,T,D} = CachedBasis{
+    𝔽,DefaultOrthonormalBasis{𝔽,TangentSpaceType},HOSVDBasis{T,D}
+}
+
+⊗ᴿ(a...) = Kronecker.:⊗(reverse(a)...)
+
+Base.:*(s::Number, x::TuckerTVector) = TuckerTVector(s * x.Ċ, s .* x.U̇)
+Base.:*(x::TuckerTVector, s::Number) = TuckerTVector(x.Ċ * s, x.U̇ .* s)
+Base.:/(x::TuckerTVector, s::Number) = TuckerTVector(x.Ċ / s, x.U̇ ./ s)
+Base.:\(s::Number, x::TuckerTVector) = TuckerTVector(s \ x.Ċ, s .\ x.U̇)
+Base.:+(x::TuckerTVector, y::TuckerTVector) = TuckerTVector(x.Ċ + y.Ċ, x.U̇ .+ y.U̇)
+Base.:-(x::TuckerTVector, y::TuckerTVector) = TuckerTVector(x.Ċ - y.Ċ, x.U̇ .- y.U̇)
+Base.:-(x::TuckerTVector) = TuckerTVector(-x.Ċ, map(-, x.U̇))
+Base.:+(x::TuckerTVector) = TuckerTVector(x.Ċ, x.U̇)
+Base.:(==)(x::TuckerTVector, y::TuckerTVector) = (x.Ċ == y.Ċ) && all(x.U̇ .== y.U̇)
```

Comment:
We should also have the broadcasting thing for `TuckerTVector`. That may be a bit more tricky than for `UMVTVector` so I can write that.

Reply:
here is the code for broadcasting:
```julia
# Tuple-like broadcasting of TuckerTVector

function Broadcast.BroadcastStyle(::Type{TuckerTVector{T,D}}) where {T,D}
    return Broadcast.Style{TuckerTVector{Any,D}}()
end
function Broadcast.BroadcastStyle(
    ::Broadcast.AbstractArrayStyle{0},
    b::Broadcast.Style{<:TuckerTVector},
)
    return b
end

Broadcast.instantiate(bc::Broadcast.Broadcasted{Broadcast.Style{<:TuckerTVector},Nothing}) = bc
function Broadcast.instantiate(bc::Broadcast.Broadcasted{Broadcast.Style{<:TuckerTVector}})
    Broadcast.check_broadcast_axes(bc.axes, bc.args...)
    return bc
end

Broadcast.broadcastable(v::TuckerTVector) = v

@inline function Base.copy(bc::Broadcast.Broadcasted{Broadcast.Style{TuckerTVector{Any,D}}}) where {D}
    return TuckerTVector(
        @inbounds(Broadcast._broadcast_getindex(bc, Val(:Ċ))),
        ntuple(i -> @inbounds(Broadcast._broadcast_getindex(bc, Val((:U̇, i)))), Val(D)),
    )
end

Base.@propagate_inbounds function Broadcast._broadcast_getindex(
    v::TuckerTVector,
    ::Val{I},
) where {I}
    if I isa Symbol
        return getfield(v, I)
    else
        return getfield(v, I[1])[I[2]]
    end
end

Base.axes(::TuckerTVector) = ()

@inline function Base.copyto!(
    dest::TuckerTVector,
    bc::Broadcast.Broadcasted{Broadcast.Style{TuckerTVector{Any,D}}},
) where {D}
    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match
    if bc.f === identity && bc.args isa Tuple{TuckerTVector} # only a single input argument to broadcast!
        A = bc.args[1]
        return copyto!(dest, A)
    end
    bc′ = Broadcast.preprocess(dest, bc)
    copyto!(dest.U, Broadcast._broadcast_getindex(bc′, Val(:Ċ)))
    for i in 1:D
        copyto!(dest.M, Broadcast._broadcast_getindex(bc, Val((:U̇, i))))
    end
    return dest
end

####
```